cmd/api/pagination/pagination.go

```diff
@@ -0,0 +1,104 @@
+	continueKey     = "continue"
+	continueDateKey = "continueDate"
+	continueUUIDKey = "continueUUID"
+)
+
+// Helper function for routes to retrieve the information needed
+// This is kept here so its close to the function that sets these
+// values in the context (Middleware)
+func GetValuesFromContext(context *gin.Context) (string, string, string) {
+	return context.GetString(limitKey), context.GetString(continueUUIDKey), context.GetString(continueDateKey)
+}
+
+func CreateToken(uuid, date string) string {
+	// The date is returned as a quoted string, so remove them quotes
+	date = strings.TrimPrefix(date, "\"")
+	date = strings.TrimSuffix(date, "\"")
+	tokenString := fmt.Sprintf("%s %s", uuid, date)
+	return base64.StdEncoding.EncodeToString([]byte(tokenString))
+}
+
+// This middleware validates the `limit` and `continue` query parameters
+// and populates `limit` and `continueValue` in the context with their
+// respective values so they are retrieved by the endpoints that need it
+func Middleware() gin.HandlerFunc {
+	return func(context *gin.Context) {
+		limitString := context.Query(limitKey)
+		continueToken := context.Query(continueKey)
+
+		var limit string
+		if limitString != "" {
+			var err error
+			limitInteger, err := strconv.Atoi(limitString)
+			if err != nil {
+				abort.Abort(context, fmt.Sprintf("Limit '%s' could not be converted to integer.", limitString), http.StatusBadRequest)
+				return
+			}
+
+			// We reserialize to avoid SQL injection. There is the possibility the
+			// value is a valid integer, but in SQL does something else.
+			limit = strconv.Itoa(limitInteger)
+		}
+
+		var continueDate string
+		var continueUUID string
+		if continueToken != "" {
+			continueBytes, err := base64.StdEncoding.DecodeString(continueToken)
+			if err != nil {
+				abort.Abort(context, "Could not decode the continuation token.", http.StatusBadRequest)
+				return
+			}
+
+			continueParts := strings.Split(string(continueBytes), " ")
+			if len(continueParts) != 2 {
+				abort.Abort(context, "Expected two elements on the continue token, received a different amount.", http.StatusBadRequest)
+				return
+			}
+
+			continueUUID = continueParts[0]
+			err = uuid.Validate(continueUUID)
+			if err != nil {
+				abort.Abort(context, "First element of the continue token is not a valid UUID", http.StatusBadRequest)
+				return
+			}
+
+			continueDate = continueParts[1]
+			continueTimestamp, err := time.Parse(time.RFC3339, continueDate)
+			if err != nil {
+				log.Printf("Error: %s", err)
+				abort.Abort(context, fmt.Sprintf("Continuation token decoded value '%s' is not valid. It should match '%s'.", continueDate, time.RFC3339), http.StatusBadRequest)
+				return
+			}
+
+			// We reserialize to avoid SQL injection. There is the possibility the
+			// value is a valid date, but in SQL does something else.
+			continueDate = continueTimestamp.Format(time.RFC3339)
+		}
+
+		// Pass the values using the context, even if not recommended
```

Comment:
I also have a blog post that says that is a bad idea: https://www.calhoun.io/pitfalls-of-context-values-and-how-to-avoid-or-mitigate-them/

My feeling is that this values are magic. You set them from a place and them retrieve them in another place. Its kind of a global value, even if scoped to the request itself. This is why I added the function that retrieves them at the top of this file: it can avoid drifting, both use the same set of constants so for a change the chances of mismatching context value names is lower, etc.

I will remove the "not recommended" as there are multiple points of view.

Reply:
I believe both blog posts say the same (I wasn't against the idea of "using context isn't recommended", just wanted to make sure that our reasons to backup that were the same)

My post:
> With context.Context, function signature becomes simpler and consistent. Not only is it easier to read, we can also write middlewares to add more functionalities
> However, such convenience doesn't come with zero downside. Firstly, the dependency and input of a function become hidden
> Another compromise when using Go context is the missing of type information.

Your post:
> Downsides to using context values
> Required data for functions is obscured
> We lose type safety at compile time

So we were exactly on the same page.

However, I believe it's good to remove the comment because we are using gin as our API framework and with gin middleware the only way to share information with the routes is to do it through the context.
Let me share here [another blog post](https://app.studyraid.com/en/read/5926/130187/gin-context#best-practices-for-using-gin-context) that gathers some best practices of using gin context, among them:

> Use context values judiciously: While it's convenient to store data in the context, avoid overusing this feature. Reserve it for request-scoped data that needs to be passed between middleware and handlers.

And this is our case, AFAICS, all the variables that we are storing in the context are request specific, in this, and in the other middleware functions that we have, so I believe we are good to go with it.